### PR TITLE
LAS 3.0 : Rearrange reader branch: Change find_sections to use byte string

### PR DESCRIPTION
**Pull request by dcslagel**
This pull-request is targeted for the rearrange-reader branch. It changes reader.py::find_sections_in_file().

- The find_sections_in_file() interface is changed to take a file_ref instead of file_obj.
- The return object has the same format except it returns a binary string instead of string for each section header.
- In las.py decode section_title from binary string to section_title_str for the expected functionality.

After this change all tests run with reasonable speed for development purposes.
29 test fail as follows:
```
FAILED tests/test_encoding.py::test_utf16lebom_cchardet
FAILED tests/test_encoding.py::test_utf16le_cchardet_fails
FAILED tests/test_encoding.py::test_utf16bebom_cchardet
FAILED tests/test_encoding.py::test_pathlib_utf16lebom_cchardet
FAILED tests/test_encoding.py::test_pathlib_utf16le_cchardet_fails
FAILED tests/test_encoding.py::test_pathlib_utf16bebom_cchardet
FAILED tests/test_examples.py::test_github
FAILED tests/test_json.py::test_json_null
FAILED tests/test_null_policy.py::test_null_policy_NULL_none
FAILED tests/test_null_policy.py::test_null_policy_custom_2
FAILED tests/test_null_policy.py::test_null_policy_runon_replaced_1
FAILED tests/test_null_policy.py::test_null_policy_runon_replaced_2
FAILED tests/test_null_policy.py::test_null_policy_runon_ok_1
FAILED tests/test_null_policy.py::test_null_policy_runon_ok_2
FAILED tests/test_open_file.py::test_open_url
FAILED tests/test_open_file.py::test_open_file_object
FAILED tests/test_open_file.py::test_open_string
FAILED tests/test_write.py::test_write_sect_widths_12
FAILED tests/test_write.py::test_write_sect_widths_20_narrow
FAILED tests/test_write.py::test_write_sect_widths_20_wide
FAILED tests/test_write.py::test_df_curve_addition_on_export
FAILED tests/test_write.py::test_multi_curve_mnemonics_rewrite
FAILED tests/test_write.py::test_multi_curve_missing_mnemonics_rewrite
FAILED tests/test_write.py::test_write_units
FAILED tests/test_write.py::test_rename_and_write_curve_mnemonic
FAILED tests/test_write.py::test_write_single_step
FAILED tests/test_write.py::test_write_12_to_20_ver_in_mem_is_12
FAILED tests/test_write.py::test_write_12_to_20_ver
FAILED tests/test_write.py::test_write_20_to_12_ver_in_mem_is_20
```

Let me know if this change could be accepted (or rejected) or
needs some additional changes before being approved and merged.

Thank you,
DC

### CODE REVIEW

**Comment by dcslagel** [comment]
I have a doubt about this implementation so putting in to hold with draft mode while continuing to look into it.

**Comment by dcslagel** [comment]
Commit 655c207 is tested on Python 2.7 and Python 3.7  It runs reasonably quickly on both.
31 Python 2.7 tests fail
25 Python 3.7 tests fail

Here are the 25 failures:
```
FAILED tests/test_encoding.py::test_utf16le_cchardet_fails
FAILED tests/test_encoding.py::test_pathlib_utf16le_cchardet_fails
FAILED tests/test_examples.py::test_github
FAILED tests/test_json.py::test_json_null
FAILED tests/test_null_policy.py::test_null_policy_NULL_none
FAILED tests/test_null_policy.py::test_null_policy_custom_2
FAILED tests/test_null_policy.py::test_null_policy_runon_replaced_1
FAILED tests/test_null_policy.py::test_null_policy_runon_replaced_2
FAILED tests/test_null_policy.py::test_null_policy_runon_ok_1
FAILED tests/test_null_policy.py::test_null_policy_runon_ok_2
FAILED tests/test_open_file.py::test_open_url
FAILED tests/test_open_file.py::test_open_file_object
FAILED tests/test_open_file.py::test_open_string
FAILED tests/test_write.py::test_write_sect_widths_12
FAILED tests/test_write.py::test_write_sect_widths_20_narrow
FAILED tests/test_write.py::test_write_sect_widths_20_wide
FAILED tests/test_write.py::test_df_curve_addition_on_export
FAILED tests/test_write.py::test_multi_curve_mnemonics_rewrite
FAILED tests/test_write.py::test_multi_curve_missing_mnemonics_rewrite
FAILED tests/test_write.py::test_write_units
FAILED tests/test_write.py::test_rename_and_write_curve_mnemonic
FAILED tests/test_write.py::test_write_single_step
FAILED tests/test_write.py::test_write_12_to_20_ver_in_mem_is_12
FAILED tests/test_write.py::test_write_12_to_20_ver
FAILED tests/test_write.py::test_write_21_to_12_ver_in_mem_is_20
```

File: lasio/las.py [hunk 1]
```diff
@@ -139,25 +137,43 @@ def read(
...
-                logger.debug(
-                    "Parsing {typ} section at lines {first_line}-{last_line} ({k} bytes) {title}".format(
-                        typ=section_type,
-                        title=section_title,
-                        first_line=first_line + 1,
-                        last_line=last_line + 1,
-                        k=k,
+
+                # Translate byte-string to string
+                section_title_str = section_title.decode(self.encoding)
```
Comment (dcslagel) [inline lasio/las.py:153]: This ```decode()``` seems like it should be in ```reader.find_sections_in_file()``` but currently we don't have the encoding available in ```reader.find_sections_in_file()```.  So if we wanted to move the ```decode()```, we would have to pass the encoding to ```reader.find_sections_in_file()```.  Let me know if that is this should be changed.

File: lasio/reader.py [hunk 2]
```diff
@@ -855,7 +919,7 @@ def read_header_line(line, pattern=None, section_name=None):
         containing a string as value.
 
     """
-    d = {"name": "", "unit": "", "value": "", "descr": ""}
+    line_data = {"name": "", "unit": "", "value": "", "descr": ""}
```
Comment (dcslagel) [inline lasio/reader.py:922]: I found ```d``` hard to search for so giving it a more identifiable name.

**Comment by dcslagel** [comment]
I moved this out of 'draft' mode so that it is in a consistent state with #358 since they are alternative solutions for find_sections().   With transform 2020 active this week,  this can wait some time for review.

Thanks,
DC

**Comment by dcslagel** [comment]
Hi @kinverarity1, 

Just following up this pull-request and the alternative approach in #358.  Is either of these proposed solutions acceptable for merge into the rearrange-reader branch?

Thank you,
DC

**Comment by kinverarity1** [comment]
Hey DC sorry about the delay with reviewing this; I'll try to get to it this week. Thanks for investigating this 👍

**Comment by dcslagel** [comment]
Thanks for the update.  :-)

**Comment by kinverarity1** [comment]
Thanks so much @dcslagel  for these PRs. This has been a blocker for me for ages now. I think we should go with your work in #358 rather than here, as it seems simpler - do you agree? If so I will merge #358 and close this PR.

**Comment by dcslagel** [comment]
Hi @kinverarity1,  

Yes,  I agree with you, #358 is clearer and will be more manageable to maintain.  Thank you for making the time to review these!

DC

**Pull request closed without merging**